File: src/main/java/dev/obscuria/obscure_tooltips/client/tooltip/element/frame/NineSlicedFrame.java
package dev.obscuria.obscure_tooltips.client.tooltip.element.frame;

import com.github.bsideup.jabel.Desugar;
import com.google.gson.JsonObject;
import dev.obscuria.obscure_tooltips.client.render.GuiGraphics;
import net.minecraft.util.JsonUtils;
import net.minecraft.util.ResourceLocation;

@Desugar
public record NineSlicedFrame(ResourceLocation textureSheet) implements TooltipFrame {

    public static NineSlicedFrame fromJson(JsonObject json) {
        return new NineSlicedFrame(new ResourceLocation(JsonUtils.getString(json, "texture_sheet")));
    }

    @Override
    public void render(GuiGraphics graphics, int x, int y, int width, int height) {
        graphics.blit(textureSheet, x - 31, y - 31, 1.0f, 1.0f, 45, 45, 140, 140);
        graphics.blit(textureSheet, x + width - 14, y - 31, 93.0f, 1.0f, 45, 45, 140, 140);
        graphics.blit(textureSheet, x - 31, y + height - 14, 1.0f, 93.0f, 45, 45, 140, 140);
        graphics.blit(textureSheet, x + width - 14, y + height - 14, 93.0f, 93.0f, 45, 45, 140, 140);
        graphics.blit(textureSheet, x + width / 2 - 23, y - 31, 47.0f, 1.0f, 45, 45, 140, 140);
        graphics.blit(textureSheet, x + width / 2 - 23, y + height - 14, 47.0f, 93.0f, 45, 45, 140, 140);
    }
}

File: common/src/main/kotlin/dev/obscuria/tooltips/client/element/frame/NineSlicedFrame.kt
package dev.obscuria.tooltips.client.element.frame

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.client.gui.GuiGraphics
import net.minecraft.resources.ResourceLocation

class NineSlicedFrame(
    val textureSheet: ResourceLocation
): ITooltipFrame {

    override fun codec(): Codec<out ITooltipFrame> = CODEC

    override fun render(graphics: GuiGraphics, x: Int, y: Int, width: Int, height: Int) {

        graphics.blit(textureSheet, x - 31, y - 31, 1f, 1f, 45, 45, 140, 140)
        graphics.blit(textureSheet, x + width - 14, y - 31, 93f, 1f, 45, 45, 140, 140)
        graphics.blit(textureSheet, x - 31, y + height - 14, 1f, 93f, 45, 45, 140, 140)
        graphics.blit(textureSheet, x + width - 14, y + height - 14, 93f, 93f, 45, 45, 140, 140)

        graphics.blit(textureSheet, x + width / 2 - 23, y - 31, 47f, 1f, 45, 45, 140, 140)
        graphics.blit(textureSheet, x + width / 2 - 23, y + height - 14, 47f, 93f, 45, 45, 140, 140)
    }

    companion object {

        val CODEC: Codec<NineSlicedFrame> = RecordCodecBuilder.create { codec ->
            codec.group(
                ResourceLocation.CODEC
                    .fieldOf("texture_sheet")
                    .forGetter(NineSlicedFrame::textureSheet)
            ).apply(codec, ::NineSlicedFrame)
        }
    }
}